layer1/chain/src/chain_sc.mligo

```diff
@@ -3,24 +3,80 @@
+    | None -> 
+        // could not find proposer, but we don't want to fail finalization, so just return no op (and dont fail)
+        []
+    | Some winner_contract -> 
+        // reward         
+        [Tezos.transaction () chain.bond_amount winner_contract]
+
+
+
+let apply_finalize (store : chain_storage) : operation list * chain_storage = 
+    match get_finalization_candidate store with
+    | Error e -> 
+        failwith ("Error during finalization:" ^ (pp_chain_error e))
+    | Ok candidate -> 
+            if is_old_enough (candidate, store, Tezos.now) then
+                reward (candidate, store), finalize (candidate, store)
+            else 
+                failwith "Error during finalization: finality period not finished"
+
+let apply_receive (proposal, store : block_proposal * chain_storage) : operation list * chain_storage =
+        // recolt bond
+        if Tezos.amount < store.bond_amount then 
+            failwith "not enough bond"
+        else
+        // store block
         begin 
-        match store_block (b, store) with
-        | Error _ -> failwith "could not store"
-        | Ok c -> [] , c
+            let new_store = increase_index store in
+            let block = make_block (proposal, new_store.max_index, Tezos.source, Tezos.now) in
+            match store_block (block, new_store) with
+            | Error _ -> 
+                failwith "could not store"
+            | Ok c -> 
+                 ([] : operation list) , c 
         end
-    | Remove i -> [] , remove_block (i, store)
 
+(* removes a block, making sure to compensate every body who proposed a block based on the removed on (recursively) 
+   /!\ this is for test. Depending on the situation, a proper removal (after refutation) might use a more complexe reimbursement strategy
+*)
+let apply_remove (i, store : index * chain_storage) =
+    let rec reward_deleted (ops, blocks, store : operation list * block list * chain_storage) : operation list = 
+        match blocks with
+        | [] -> ops
+        | b :: q -> 
+            let reward_ops = reward (b, store) in
+            let new_ops = Stdlib_ext.ListExt.concat reward_ops ops in
+            reward_deleted (new_ops, q, store)
+    in
+    let blocks, chain = remove_block (i, store) in
+    let ops = reward_deleted (([] : operation list) , blocks, chain) in
+    ops, chain
+
+(* ENDPOINTS *)
+let main (action, store : chain_parameter * chain_storage) : operation list * chain_storage = 
+    match action with
+    | Receive b -> apply_receive (b, store)
+    | Finalize -> apply_finalize store
+    | Remove i -> apply_remove (i, store)
 
-(* VIEWS *)
+(* VIEWS *) 
 
 [@view] let get_chain ((),s: unit * chain_storage) : chain_storage = s
-[@view] let get_latest ((),s: unit * chain_storage) : block option = find_latest_existing s
\ No newline at end of file
+[@view] let get_latest ((),s: unit * chain_storage) : block option = find_latest_existing s
```

Comment:
I missed this before.
`chain_storage` includes `big_map`. It's forbidden for views.

```
File "chain/src/chain_sc.mligo", line 25, characters 23-27:

Invalid view.
Type 'record[blocks -> big_map (nat , record[hash -> bytes , level -> nat , parent -> nat , proposer -> address]) , children -> big_map (nat , list (nat)) , max_index -> nat]' is forbidden as output argument.
```

Reply:
> I missed this before. `chain_storage` includes `big_map`. It's forbidden in input and output type for views.
> 
> ```
> File "chain/src/chain_sc.mligo", line 25, characters 23-27:
> 
> Invalid view.
> Type 'record[blocks -> big_map (nat , record[hash -> bytes , level -> nat , parent -> nat , proposer -> address]) , children -> big_map (nat , list (nat)) , max_index -> nat]' is forbidden as output argument.
> ```

I clearly didn't take the views seriously enough. So there is no way to add views to a contract with `big_map` storage ? I didn't expect that.

Reply:
oh, I misunderstand about ligo.
so if the type of views is `arg * storage -> return`
`arg` and `return` can't have `big_map` but in `storage` is ok.
you can operate `big_map` in views, just can't return `big_map`